src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/Error/SqlExtendedQueryTagErrorStoreV4.cs

```diff
@@ -0,0 +1,117 @@
+
+        protected SqlConnectionWrapperFactory ConnectionWrapperFactory { get; }
+
+        protected ILogger Logger { get; }
+
+        public override async Task<int> AddExtendedQueryTagErrorAsync(
+            int tagKey,
+            short errorCode,
+            long watermark,
+            DateTime createdTime,
+            CancellationToken cancellationToken = default)
+        {
+            using SqlConnectionWrapper sqlConnectionWrapper = await ConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken);
+            using SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand();
+            VLatest.AddExtendedQueryTagError.PopulateCommand(
+                sqlCommandWrapper,
+                tagKey,
+                errorCode,
+                watermark,
+                createdTime);
+
+            try
+            {
+                return (int)await sqlCommandWrapper.ExecuteScalarAsync(cancellationToken);
+            }
+            catch (SqlException e)
+            {
+                if (e.Number == SqlErrorCodes.NotFound)
+                {
+                    throw new ExtendedQueryTagNotFoundException(
+                        string.Format(
+                            CultureInfo.InvariantCulture,
+                            DicomSqlServerResource.ExtendedQueryTagNotFoundWhenAddingError,
+                            tagKey));
+                }
+
+                throw new DataStoreException(e);
+            }
+        }
+
+        public override async Task<IReadOnlyList<ExtendedQueryTagError>> GetExtendedQueryTagErrorsAsync(string tagPath, CancellationToken cancellationToken = default)
+        {
+            List<ExtendedQueryTagError> results = new List<ExtendedQueryTagError>();
+
+            using SqlConnectionWrapper sqlConnectionWrapper = await ConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken);
+            using SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand();
+
+            VLatest.GetExtendedQueryTagErrors.PopulateCommand(sqlCommandWrapper, tagPath);
+
+            try
+            {
+                using SqlDataReader reader = await sqlCommandWrapper.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken);
+                while (await reader.ReadAsync(cancellationToken))
+                {
+                    (int tagkey, int errorCode, DateTime createdTime, string studyInstanceUid, string seriesInstanceUid, string sopInstanceUid) = reader.ReadRow(
+                        VLatest.ExtendedQueryTagError.TagKey,
+                        VLatest.ExtendedQueryTagError.ErrorCode,
+                        VLatest.ExtendedQueryTagError.CreatedTime,
+                        VLatest.Instance.StudyInstanceUid,
+                        VLatest.Instance.SeriesInstanceUid,
+                        VLatest.Instance.SopInstanceUid);
+
+                    //TODO: build the error message here
```

Comment:
When will this be fixed. If we are converting error code to message here? Why can convert otherway on the way in? Also all validation error types will need a strongly typed error code, if we follow this, otherwise there is a chance of a new error message without error code falling off

Reply:
for the sql part, having an error code would mean less storage. The error types are created in Peng's PR (https://github.com/microsoft/dicom-server/pull/911), once that one is merged, I can work on the conversion.